File: July2021/01Matrix.java
class Solution {

  // TC : O(n*m)
  public int[][] updateMatrix(int[][] matrix) {
    Queue<int[]> queue=new LinkedList<>();
    for(int i=0;i<matrix.length;i++){
      for(int j=0;j<matrix[0].length;j++){
        if(matrix[i][j]==0)
          queue.add(new int[]{i,j});
        else
          matrix[i][j]=-1;
      }
    }
    int[][] dirs= {{0,1},{-1,0},{1,0},{0,-1}};

    int length=0;
    while(!queue.isEmpty()){
      int size=queue.size();
      length++;
      while(size-->0){
        int[] cell=queue.poll();
        for(int[] dir:dirs){
          int r=cell[0]+dir[0];
          int c=cell[1]+dir[1];
          if(r<0 || c<0 || r==matrix.length || c==matrix[0].length || matrix[r][c]!=-1)                                   continue;
          matrix[r][c]=length;
          queue.add(new int[]{r,c});
        }
      }
    }
    return matrix;
  }
}

File: August2023 /01 Matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        queue<pair<int,int>> q;
        vector<vector<bool>> visited(mat.size(),vector<bool>(mat[0].size()));
        for(int i=0;i<mat.size();i++){
            for(int j=0;j<mat[0].size();j++){
                if(!mat[i][j]) q.push({i,j});
            }
        }
        vector<vector<int>> dir={{0,1},{0,-1},{1,0},{-1,0}};
        int level=0;
        while(q.size())
        {
            int s=q.size();
            for(int i=0;i<s;i++)
            {
                int row=q.front().first;
                int col=q.front().second;
                mat[row][col]=level;
                q.pop();
                for(auto& i:dir){
                    row+=i[0];
                    col+=i[1];
                    if(!(col<0 || row<0 || col>=mat[0].size() || row>=mat.size() ||visited[row][col] || mat[row][col]==0))
                    {
                        visited[row][col]=true;
                        q.push({row,col});
                    }
                    row-=i[0];
                    col-=i[1];
                } 

            }
            level++;
        }
        return mat;
    }
};